### PR TITLE
feat(gatsby-transformer-remark, gatsby-plugin-mdx) Make timeToRead configurable

### PR DESCRIPTION
**Pull request by johnrichter**
I've been wanting this for a while (since not everyone reads at 265 wpm) and had some time to work on it this evening.

## Description

This PR makes `timeToRead` calculation user configurable. I went with the simple solution to start.

```js
{ 
  // Mimics a reading rate of 42 words per minute
  timeToRead: (wordCount, html, md) => wordCount / 42
}
```

### More advanced potential solution

It would make sense to give the user more access to the markdown node for more flexible and precise calculation (e.g. code samples and complex graphs). I was thinking this could look like 

```js
{
  timeToRead: (wordCount, { markdownNode, getAST, getMarkdownAST, getHeadings, getTableOfContents, getHTMLAst, getHTML, getExcerptAst, getExcerpt }) => wordCount / 42
}
```

 It essentially gives access to all of the `markdownNode` field resolvers. `wordCount` would now be based off of `markdownNode.rawMarkdownBody` instead of the sanitized HTML. It's more complicated, but gives the user everything they'd need to calculate `timeToRead` as precisely as they'd want.

## Related Issues

Related to https://github.com/gatsbyjs/gatsby/issues/16760. It would allow users to calculate the value in the they want to vs Gatsby building and maintaining a custom one-size-fits-all solution.

## `gatsby-plugin-mdx`

This plugin also has a `timeToRead` that can be updated with the same simple solution. It doesn't look on first glance that it could be updated to support the more advanced solution above though.

### CODE REVIEW

**Comment by johnrichter** [comment]
I went ahead and added support to `gatsby-plugin-mdx` as well. I could use some help writing tests for it though. I have no idea what I'm doing when it comes up to setting up brand new tests for a gatsby plugin (or a file within that plugin) and I'd like to make sure a test exists before it gets shipped.

**Comment by johnrichter** [comment]
That is a good question. My guess would be that it isn't something that a lot of people care about. Searching `timeToRead` in the Gatsby repo only returns a few issues that mention it specifically. The ones that do exist are about the accuracy of the count when considering things like code blocks. 

What I do think could be improved more generally with the default calculation is switching away from using the HTML string to something without syntax of any kind. It should be possible to parse Markdown AST to do this in a fairly straightforward manner. If you look at [this line](https://github.com/gatsbyjs/gatsby/pull/19763/files#diff-1a32d7dae37831084eef6051a845b15fR898) of the PR, the expected count should be `7`, but it is `9` because the paragraph tags are parsed as words. This is also happening in the default case and I suspect part of the reason why the default WPM is quite high (imo).

**Comment by pieh** [comment]
We are doing some pure text extraction / stripping tags for `excerpt` so maybe something that could be reused?

**Comment by pieh** [comment]
Also, if current `const avgWPM = 265` is unusually high, maybe we should adjust that as well? I'm not sure where that value came from honestly (I can trace it back to https://github.com/gatsbyjs/gatsby/commit/509175e4dadf3a99476edab1a477499d4fa8bcb1#diff-db344a7cdc7479489764106f94b249a0R113 which seems like initial commit for what now is `gatsby-transformer-remark` 3 years ago and it wasn't touched since then)

**Comment by pvdz** [comment]
@johnrichter @pieh what do we want to do with this?

John, the changes you suggested sound good in any case, do you have space to try and improve that system?

**Comment by johnrichter** [comment]
I have a bit of time next week to try to get a more accurate solution in place. 

@pieh's idea of adapting the plain excerpt implementation for this is a great starting point. I'd like to keep an escape hatch for those that would want to customize the calculation with the goal of making it useful for edge cases like customizing how long inline React components would take to read.

I'm thinking that I could try to do a bit of research to figure out good defaults for the the time it takes to process images and code blocks. The latter of which will be tricky since some languages are much more readable than others. [Paul Stamatiou](https://paulstamatiou.com/about-this-website/#tech) recently mentioned that he 
> read somewhere that Medium.com adds 12 seconds to the estimated reading time per image used in the article.

That makes me hopeful that I could find some sort of useful default values.

**Comment by LekoArts** [comment]
@johnrichter Hi, do you have time to wrap up this PR? :)

**Comment by johnrichter** [comment]
Hi @LekoArts, at the moment I don't have the time to go into the level of detail I mentioned above. If the changes in this PR are good enough for now I have enough spare cycles to rebase and complete whatever else is needed to get this shipped. My apologies for not getting back to y'all on this PR sooner!

I still want to improve the calculation in a more meaningful way in the future, but if you're looking at officially overhauling it by all means go for it 🙂

File: packages/gatsby-plugin-mdx/README.md [hunk 1]
```diff
@@ -417,6 +418,27 @@ Gatsby includes the media-type of the content on any given node. For
...
+rate or by custom heuristics based on the markdown node.
+
+```js
+// gatsby-config.js
+module.exports = {
+  plugins: [
+    {
+      resolve: `gatsby-plugin-mdx`,
+      options: {
+        timeToRead: (wordCount, mdxNode) => wordCount / 42,
```
Comment (muescha) [inline packages/gatsby-plugin-mdx/README.md]: not match `timeToRead(wordCount, html, md)`

File: packages/gatsby-plugin-mdx/gatsby/source-nodes.js [hunk 2]
```diff
@@ -221,13 +221,17 @@ ${e}`
...
-          timeToRead = Math.round(words / avgWPM)
-          if (timeToRead === 0) {
-            timeToRead = 1
-          }
-          return timeToRead
+          const timeToReadInMinutes = Math.max(
+            1,
+            Math.round(
+              _.isFunction(timeToRead)
+                ? timeToRead(words, mdxNode)
```
Comment (muescha) [inline packages/gatsby-plugin-mdx/gatsby/source-nodes.js]: not match `timeToRead(wordCount, html, md)`

Comment (johnrichter) [reply packages/gatsby-plugin-mdx/gatsby/source-nodes.js]: 👋 I had to add some functionality, but the `timeToRead` signature now matches for both MDX and Remark.

File: packages/gatsby-plugin-mdx/README.md (hunk 1 above)
Comment (johnrichter) [reply packages/gatsby-plugin-mdx/README.md]: Updated 👍

**Comment by johnrichter** [comment]
I've rebased on master as well

File: packages/gatsby-transformer-remark/README.md [hunk 3]
```diff
@@ -22,6 +22,8 @@ plugins: [
       pedantic: true,
       // GitHub Flavored Markdown mode (default: true)
       gfm: true,
+      // Calculate timeToRead in minutes using word count, sanitized html, and raw Markdown content. (default: wordCount / 265)
+      timeToRead: (wordCount, html, rawMD) => wordCount / 42
```
Comment (muescha) [inline packages/gatsby-transformer-remark/README.md]:
```suggestion
      timeToRead: (wordCount, html, rawMDX) => wordCount / 42
```

File: packages/gatsby-transformer-remark/src/extend-node-type.js [hunk 4]
```diff
@@ -604,7 +605,9 @@ module.exports = (
       timeToRead: {
         type: `Int`,
         resolve(markdownNode) {
-          return getHTML(markdownNode).then(timeToRead)
+          return getHTML(markdownNode).then(html =>
+            timeToRead(markdownNode, html, calcTimeToRead)
```
Comment (muescha) [inline packages/gatsby-transformer-remark/src/extend-node-type.js]: it is more easy to understand if you rename the utils function
```suggestion
            timeToReadCalculator(markdownNode, html, timeToRead)
```

**Comment by muescha** [comment]
just now the function gets different TTR? in `-mdx` you don't count the non latin chars

File: packages/gatsby-transformer-remark/src/extend-node-type.js (hunk 4 above)
Comment (johnrichter) [reply packages/gatsby-transformer-remark/src/extend-node-type.js]: That is a great point. I thought it was confusing too to be honest and didn't think about renaming the util function. Let me know if you like the new naming 👍

**Comment by johnrichter** [comment]
For the original code in the PR, the utility functions hadn't been created yet and all of that was mostly inline. The utility function still resides in the Remark transformer so I don't want to create a dependency on the Remark transformer in the MDX plugin. I could extract the utils to a shared package somewhere (optimal) or just copy and paste them to mdx (have to maintain in both places). Which one depends on what y'all would like to do really since I'm indifferent.

Historically, MDX and Remark have had different TTRs so I left it as is. For a first pass at creating better time to read calculations the main focus of this PR is passing the responsibility to the user if they want to change it.

**Comment by muescha** [comment]
yes you are right, that is not the scope of this PR

**Comment by pvdz** [comment]
@johnrichter tests are failing even after rebase. If we can get the tests pass I'm inclined to merge it. Please let me know whether you want to fix the tests.

**Comment by johnrichter** [comment]
Hi @pvdz, I'm working on getting those tests working today. Did a quick rebase on master again before I start working on getting the tests to pass.

**Comment by johnrichter** [comment]
@pvdz Looks like rebasing was all that was needed 👍

**Comment by pvdz** [comment]
@johnrichter Thank you for the additional effort :purple_heart:  Sorry it took this long, but happy we can merge it now :D

**Comment by johnrichter** [comment]
You're welcome! Thanks for merging it in. I can't wait to use it when it gets released 😄 

Also, I had no idea ya'll hand out swag for PRs 😍 😍 😍

**Comment by wardpeet** [comment]
@johnrichter we had to revert this change because of Out of memory issues and slow query builds.

see https://github.com/gatsbyjs/gatsby/issues/23614 & https://github.com/gatsbyjs/gatsby/issues/23714

**Comment by johnrichter** [comment]
☹️ Interesting. In practice all this did was add a plugin option and leave the original code intact. Perhaps the changes that were pulled in by rebasing changed something more fundamental. I'll take a look and see what happened.

**Comment by muescha** [comment]
i have two ideas:

1) maybe only use genHtml only if there is a custom function defined? other wise there no calculation needed for html
2) more general: maybe it is possible to cache some calculations for the resolvers (i see the some functions called many times for example: `genHtml`, `processMdx`) 
  i see `getAST` has a cache - maybe apply the same to the others....

**Pull request merged**